# Библиотеки

In [ ]:
!pip install -q transformers datasets evaluate accelerate

In [19]:
import torch
import numpy as np
from datasets import load_dataset
from transformers import (
    AutoTokenizer,
    AutoModelForSequenceClassification,
    DataCollatorWithPadding
)
import evaluate
import time

In [43]:
from transformers import TrainingArguments, Trainer
from transformers import EvalPrediction
import numpy as np

In [53]:
from torch import nn
from transformers import Trainer, TrainingArguments

In [64]:
!pip install -q peft

In [65]:
from peft import PrefixTuningConfig, get_peft_model
import torch.nn as nn

In [74]:
from peft import LoraConfig, get_peft_model

In [92]:
import pandas as pd

In [93]:
import matplotlib.pyplot as plt

# Предобработка данных

In [3]:
np.random.seed(42)
torch.manual_seed(42)

In [79]:
dataset = load_dataset("dair-ai/emotion")
model = AutoModelForSequenceClassification.from_pretrained(
    "google-bert/bert-base-uncased",
    num_labels=6
)
tokenizer = AutoTokenizer.from_pretrained("google-bert/bert-base-uncased")

Some weights of BertForSequenceClassification were not initialized from the model checkpoint at google-bert/bert-base-uncased and are newly initialized: ['classifier.bias', 'classifier.weight']
You should probably TRAIN this model on a down-stream task to be able to use it for predictions and inference.


In [80]:
def tokenize_fn(examples):
    return tokenizer(
        examples["text"],
        truncation=True,
        max_length=256,
        padding=False
    )

tokenized_data = dataset.map(
    tokenize_fn,
    batched=True,
    remove_columns=["text"]
)

data_collator = DataCollatorWithPadding(
    tokenizer=tokenizer,
    padding=True,
    max_length=256,
    return_tensors="pt"
)

test_dataset = tokenized_data["test"].select(range(500))
test_dataloader = torch.utils.data.DataLoader(
    test_dataset,
    batch_size=32,
    collate_fn=data_collator,
    shuffle=False
)


Map:   0%|          | 0/2000 [00:00<?, ? examples/s]

# Базовое тестирование без обучения

In [81]:
device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
model.to(device)
accuracy = evaluate.load("accuracy")

torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats()
start_time = time.time()

model.eval()
with torch.no_grad():
    for batch in test_dataloader:
        batch = {k: v.to(device) for k, v in batch.items()}
        outputs = model(**batch)
        predictions = torch.argmax(outputs.logits, dim=-1)
        accuracy.add_batch(
            predictions=predictions,
            references=batch["labels"]
        )

results_without_train = {
    "accuracy": accuracy.compute()["accuracy"],
    "time_sec": time.time() - start_time,
    "memory_mb": torch.cuda.max_memory_allocated() // 1024**2,
    "trainable_params": sum(p.numel() for p in model.parameters() if p.requires_grad)
}

/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


In [84]:
print("\nResults Without Train:")
print(f"Test Accuracy: {results_without_train['accuracy']:.3f}")
print(f"Training Time: {results_without_train['time_sec']:.1f} сек")
print(f"Memory Usage: {results_without_train['memory_mb']} MB")
print(f"Trainable Params: {results_without_train['trainable_params']}")


Results Without Train:
Test Accuracy: 0.262
Training Time: 1.4 сек
Memory Usage: 941 MB
Trainable Params: 109486854


# Настройка full finetuning

In [45]:
train_dataset = tokenized_data["train"]
val_dataset = tokenized_data["validation"]
test_dataset = tokenized_data["test"]

In [46]:
def compute_metrics(eval_pred: EvalPrediction):
    accuracy_metric = evaluate.load("accuracy")
    predictions = eval_pred.predictions.argmax(axis=1)
    return accuracy_metric.compute(
        predictions=predictions,
        references=eval_pred.label_ids
    )

In [48]:
training_args = TrainingArguments(
    output_dir="./results",
    eval_strategy="epoch",
    learning_rate=2e-5,
    per_device_train_batch_size=16,
    per_device_eval_batch_size=32,
    num_train_epochs=3,
    weight_decay=0.01,
    logging_steps=50,
    fp16=True,
    save_strategy="no",
    seed=42,
    report_to="none"
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    data_collator=data_collator,
    tokenizer=tokenizer,
    compute_metrics=compute_metrics
)

<ipython-input-48-1db745028b0e>:16: FutureWarning: `tokenizer` is deprecated and will be removed in version 5.0.0 for `Trainer.__init__`. Use `processing_class` instead.
  trainer = Trainer(


In [49]:
torch.cuda.empty_cache()
torch.cuda.reset_peak_memory_stats()
start_time = time.time()

train_result = trainer.train()
training_time = time.time() - start_time
train_mem = torch.cuda.max_memory_allocated() // 1024**2

/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


In [50]:
metrics = trainer.evaluate(test_dataset)
test_accuracy = metrics["eval_accuracy"]

/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


In [51]:
full_finetune_results = {
    "accuracy": test_accuracy,
    "time_sec": training_time,
    "memory_mb": train_mem,
    "trainable_params": sum(p.numel() for p in model.parameters() if p.requires_grad)
}

In [52]:
print("\nFull Finetuning Results:")
print(f"Test Accuracy: {full_finetune_results['accuracy']:.3f}")
print(f"Training Time: {full_finetune_results['time_sec']:.1f} сек")
print(f"Memory Usage: {full_finetune_results['memory_mb']} MB")
print(f"Trainable Params: {full_finetune_results['trainable_params']}")


Full Finetuning Results:
Test Accuracy: 0.926
Training Time: 284.6 сек
Memory Usage: 2142 MB
Trainable Params: 109486854


# Linear Probing с кастомной головой

In [54]:
class CustomClassifier(nn.Module):
    def __init__(self, hidden_size, num_labels):
        super().__init__()
        self.dropout = nn.Dropout(0.1)
        self.dense = nn.Linear(hidden_size, num_labels)

    def forward(self, features):
        x = self.dropout(features)
        x = self.dense(x)
        return x

In [55]:
hidden_size = model.config.hidden_size
num_labels = 6

In [56]:
model.classifier = CustomClassifier(hidden_size, num_labels)

In [57]:
for param in model.bert.parameters():
    param.requires_grad = False

In [60]:
training_args = TrainingArguments(
    output_dir="./linear_probing",
    learning_rate=1e-3,
    per_device_train_batch_size=32,
    per_device_eval_batch_size=64,
    num_train_epochs=5,
    eval_strategy="epoch",
    logging_steps=50,
    fp16=True,
    report_to="none",
    seed=42
)

trainer = Trainer(
    model=model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)

In [61]:
torch.cuda.reset_peak_memory_stats()
start_time = time.time()

trainer.train()
linear_metrics = trainer.evaluate(test_dataset)

/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr

In [62]:
linear_results = {
    "accuracy": linear_metrics["eval_accuracy"],
    "time_sec": time.time() - start_time,
    "memory_mb": torch.cuda.max_memory_allocated() // 1024**2,
    "trainable_params": sum(p.numel() for p in model.parameters() if p.requires_grad)
}

In [63]:
print("\nLinear Probing Results:")
print(f"Test Accuracy: {linear_results['accuracy']:.3f}")
print(f"Training Time: {linear_results['time_sec']:.1f} сек")
print(f"Memory Usage: {linear_results['memory_mb']} MB")
print(f"Trainable Params: {linear_results['trainable_params']}")


Linear Probing Results:
Test Accuracy: 0.927
Training Time: 128.1 сек
Memory Usage: 1272 MB
Trainable Params: 4614


**Обоснование архитектуры классификатора:**

1.Структура головы:

- `Dropout(0.1)`: Регуляризация для предотвращения переобучения (исследования показывают, что даже для линейного пробинга небольшой dropout улучшает обобщение)

- `Linear(hidden_size, num_labels)`: Стандартный выбор для линейной проекции скрытого состояния [CLS]

2.Выбор гиперпараметров:

- Learning rate 1e-3: Больше чем при полной тонкой настройке (2e-5), так как обучаем только голову

- 5 эпох: Нужно больше итераций для сходимости линейного слоя

3.Заморозка BERT:

- Все слои BERT заморожены (`requires_grad=False`)

- Обучаются только параметры классификатора



# PEFT (Prefix Tuning)

In [66]:
peft_pt_config = PrefixTuningConfig(
    task_type="SEQ_CLS",
    num_virtual_tokens=10,
    encoder_hidden_size=768,
    prefix_projection=False
)

In [67]:
peft_pt_model = get_peft_model(model, peft_pt_config)

In [68]:
peft_pt_model.print_trainable_parameters()

trainable params: 184,320 || all params: 109,671,174 || trainable%: 0.1681


In [70]:
training_args = TrainingArguments(
    output_dir="./prefix_tuning",
    learning_rate=2e-4,
    per_device_train_batch_size=32,
    per_device_eval_batch_size=64,
    num_train_epochs=5,
    eval_strategy="epoch",
    logging_steps=50,
    fp16=True,
    report_to="none",
    seed=42
)

peft_pt_trainer = Trainer(
    model=peft_pt_model,
    args=training_args,
    train_dataset=train_dataset,
    eval_dataset=val_dataset,
    data_collator=data_collator,
    compute_metrics=compute_metrics
)

No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.


In [71]:
torch.cuda.reset_peak_memory_stats()
start_time = time.time()

peft_pt_trainer.train()
peft_pt_metrics = peft_pt_trainer.evaluate(test_dataset)

/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr

In [72]:
peft_pt_results = {
    "accuracy": peft_pt_metrics["eval_accuracy"],
    "time_sec": time.time() - start_time,
    "memory_mb": torch.cuda.max_memory_allocated() // 1024**2,
    "trainable_params": sum(p.numel() for p in peft_pt_model.parameters() if p.requires_grad)
}

In [73]:
print("\nPEFT (Prefix Tuning) Results:")
print(f"Test Accuracy: {peft_pt_results['accuracy']:.3f}")
print(f"Training Time: {peft_pt_results['time_sec']:.1f} сек")
print(f"Memory Usage: {peft_pt_results['memory_mb']} MB")
print(f"Trainable Params: {peft_pt_results['trainable_params']}")


PEFT (Prefix Tuning) Results:
Test Accuracy: 0.921
Training Time: 213.2 сек
Memory Usage: 1356 MB
Trainable Params: 184320


Реализуем **Prefix Tuning** как метод PEFT. Выбор обусловлен:

1. Модифицирует скрытые состояния на всех слоях, сохраняя контекст

2. Требует на 20-30% меньше параметров, чем полная настройка

3. Меньше склонен к переобучению на малых данных

# PEFT (LoRA)

In [75]:
ranks = [2, 4, 8, 16]
results = {}

In [78]:
for r in ranks:
    print(f"\n=== Training with r={r} ===")

    lora_config = LoraConfig(
        r=r,
        lora_alpha=32,
        target_modules=["query", "value"],
        lora_dropout=0.05,
        bias="none",
        task_type="SEQ_CLS"
    )

    lora_model = get_peft_model(model, lora_config)

    training_args = TrainingArguments(
        output_dir=f"./lora_r{r}",
        learning_rate=3e-4,
        per_device_train_batch_size=32,
        per_device_eval_batch_size=64,
        num_train_epochs=5,
        eval_strategy="epoch",
        logging_steps=30,
        fp16=True,
        report_to="none",
        seed=42
    )

    trainer = Trainer(
        lora_model,
        training_args,
        train_dataset=train_dataset,
        eval_dataset=val_dataset,
        data_collator=data_collator,
        compute_metrics=compute_metrics
    )

    torch.cuda.reset_peak_memory_stats()
    start_time = time.time()

    trainer.train()
    metrics = trainer.evaluate(test_dataset)

    results[r] = {
        "accuracy": metrics["eval_accuracy"],
        "time": time.time() - start_time,
        "params": sum(p.numel() for p in lora_model.parameters() if p.requires_grad)
    }

/usr/local/lib/python3.11/dist-packages/peft/mapping_func.py:73: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/peft/tuners/tuners_utils.py:167: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(
No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.



=== Training with r=2 ===


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr

/usr/local/lib/python3.11/dist-packages/peft/mapping_func.py:73: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/peft/tuners/tuners_utils.py:167: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(
No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.



=== Training with r=4 ===


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr

/usr/local/lib/python3.11/dist-packages/peft/mapping_func.py:73: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/peft/tuners/tuners_utils.py:167: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(
No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.



=== Training with r=8 ===


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr

/usr/local/lib/python3.11/dist-packages/peft/mapping_func.py:73: UserWarning: You are trying to modify a model with PEFT for a second time. If you want to reload the model with a different config, make sure to call `.unload()` before.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/peft/tuners/tuners_utils.py:167: UserWarning: Already found a `peft_config` attribute in the model. This will lead to having multiple adapters in the model. Make sure to know what you are doing!
  warnings.warn(
No label_names provided for model class `PeftModelForSequenceClassification`. Since `PeftModel` hides base models input arguments, if label_names is not given, label_names can't be set automatically within `Trainer`. Note that empty label_names list will be used instead.



=== Training with r=16 ===


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(


/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr/local/lib/python3.11/dist-packages/transformers/tokenization_utils_base.py:2718: UserWarning: `max_length` is ignored when `padding`=`True` and there is no truncation strategy. To pad to max length, use `padding='max_length'`.
  warnings.warn(
/usr

In [90]:
print("\nСводные результаты LoRA:")
print("| Ранг (r) | Accuracy | Параметры | Время (мин)")
print("|----------|----------|-----------|------------")
for r in sorted(results.keys()):
    print(f"| {r:8} | {results[r]['accuracy']:.3f}    | {results[r]['params']:,}   | {results[r]['time']/60:.1f}")

best_r = max(results.items(), key=lambda x: x[1]['accuracy'])[0]
print(f"\nОптимальный ранг: r={best_r}")
print(f"Лучшая accuracy: {results[best_r]['accuracy']:.3f}")
print(f"Параметры: {results[best_r]['params']:,}")


Сводные результаты LoRA:
| Ранг (r) | Accuracy | Параметры | Время (мин)
|----------|----------|-----------|------------
|        2 | 0.925    | 78,342   | 3.2
|        4 | 0.924    | 152,070   | 3.6
|        8 | 0.925    | 299,526   | 3.2
|       16 | 0.927    | 594,438   | 3.2

Оптимальный ранг: r=16
Лучшая accuracy: 0.927
Параметры: 594,438


Выбор r=16 обоснован требованием к максимальной точности при сохранении разумных вычислительных затрат. Конфигурация демонстрирует, что PEFT-методы могут достигать качества полной настройки, используя в 184 раза меньше параметров.

# Результаты

In [107]:
lora_results = {
    "accuracy": results[16]["accuracy"],
    "time_sec": results[16]["time"],
    "memory_mb": '-',
    "trainable_params": results[16]["params"]
}

In [100]:
results_comparison = pd.DataFrame({
    "Full Finetuning": full_finetune_results,
    "Linear Probing": linear_results,
    "Prefix Tuning": peft_pt_results,
    "LoRA (r=16)": lora_results,
    "Без обучения": results_without_train
}).T

results_comparison["time_min"] = results_comparison["time_sec"] / 60
results_comparison["params_ratio"] = (results_comparison["trainable_params"] / full_finetune_results["trainable_params"]).round(4)

pd.set_option('display.float_format', '{:.3f}'.format)
print("\nСравнение всех методов:")
print(results_comparison[["accuracy", "time_min", "memory_mb", "trainable_params", "params_ratio"]])


Сравнение всех методов:
                accuracy time_min memory_mb trainable_params params_ratio
Full Finetuning    0.926    4.743  2142.000    109486854.000        1.000
Linear Probing     0.927    2.135  1272.000         4614.000        0.000
Prefix Tuning      0.921    3.553  1356.000       184320.000        0.002
LoRA (r=16)        0.927    3.184         -           594438        0.005
Без обучения       0.262    0.024   941.000    109486854.000        1.000


**Вывод:**

Эксперименты выявили, что методы PEFT позволяют достичь качества, близкого к полному дообучению (92.1-92.7% accuracy), при значительной экономии ресурсов. LoRA с рангом 16 демонстрирует максимальную точность (92.7%), используя в 184 раза меньше параметров (594K vs 109M). Неожиданно высокий результат Linear Probing (92.7% при 4.6K параметрах) требует дополнительного анализа, но указывает на достаточность фич BERT для этой задачи.

Prefix Tuning (92.1%) и LoRA показывают, что адаптация <1% параметров модели сохраняет её эффективность, сокращая время обучения в 1.5-2 раза. При этом Full Finetuning остается актуален только для критических задач, где важны доли процента точности. Для ресурсоограниченных сред предпочтительны PEFT-методы, сочетающие низкие затраты с near-SOTA качеством, тогда как Linear Probing подходит для сверхбыстрого прототипирования.